### PR TITLE
[Consensus Voting] `EventLoopV2` -> `EventLoop`

### PR DESCRIPTION
**Pull request by durkmurder**
### Context

This PR replaces event loop with new implementation and breaks lots of tests :smile: 
I have tried to keep the changes very minimal, there is some room for further changes, like consolidating `module.Hotstuff` and `module.HotstuffFollower` but not for this PR.

Notable changes:

- Event loop is broken since it doesn't have `EventHandlerV2` which is pending review at this point: https://github.com/onflow/flow-go/pull/1168
- Compliance engine is broken as well since it doesn't push votes, it will depend on vote aggregator which will be replaced as part of other PR. 
- Those are two primary reasons why many tests are disabled since they depend on compliance engine and event handler.

### What was done:

- Deleted old `EventLoop` and replaced with `EventLoopV2`
- Skipped unit tests that depend on event handler
- Skipped all integration tests since compliance engine is broken and almost every test creates a consensus node

### CODE REVIEW

File: integration/Makefile [hunk 1]
```diff
@@ -11,8 +11,11 @@ test:
 integration-test: access-tests common-tests execution-tests verification-tests collection-tests consensus-tests epoch-tests
 
 # NOTE: Currently skipping collection and consensus tests due to them not passing properly on CI
+# NOTE: Currently skipping all integration tests while replacing modules for consensus V2
 .PHONY: ci-integration-test
-ci-integration-test: access-tests common-tests execution-tests epoch-tests verification-tests # collection-tests # consensus-tests
+#ci-integration-test: access-tests common-tests execution-tests epoch-tests verification-tests # collection-tests # consensus-tests
+ci-integration-test:
+	exit 0
```
Comment (jordanschalm) [inline integration/Makefile:18]: Why do we need to disable the tests, given that this isn't being merged to a branch requiring CI, and by the time it is merged to master I assume we will have turned these back on anyway.

File: engine/collection/compliance/engine.go [hunk 2]
```diff
@@ -45,6 +47,7 @@ type Engine struct {
 	pendingVotes   engine.MessageStore
 	messageHandler *engine.MessageHandler
 	con            network.Conduit
+	stopHotstuff   context.CancelFunc
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply engine/collection/compliance/engine.go:50]: oops, that's right, didn't do this for collection engine

File: integration/Makefile (hunk 1 above)
Comment (durkmurder) [reply integration/Makefile:18]: that's eventually merged in `feature/consensus-voting`, I've created it this way to make the diff smaller.

**Review by jordanschalm — APPROVED** [review]

File: consensus/hotstuff/event_loop.go [hunk 3]
```diff
@@ -1,216 +1,13 @@
...
-
-// Done implements interface module.ReadyDoneAware
-func (el *EventLoop) Done() <-chan struct{} {
-	el.lm.OnStop(func() {
-		// wait for event loop to exit
-		<-el.unit.Done()
-	})
-	return el.lm.Stopped()
+	// SubmitTrustedQC accepts QC for processing. QC will be dispatched on worker thread.
+	// CAUTION: QC is trusted (_not_ validated again)
```
Comment (zhangchiqing) [inline consensus/hotstuff/event_loop.go:11]:
```suggestion
	// CAUTION: QC is trusted (_not_ validated again), as it's built by ourselves.
```

File: engine/consensus/compliance/engine.go [hunk 4]
```diff
@@ -179,6 +188,7 @@ func (e *Engine) Ready() <-chan struct{} {
 func (e *Engine) Done() <-chan struct{} {
 	e.lm.OnStop(func() {
 		e.log.Debug().Msg("shutting down hotstuff eventloop")
+		e.stopHotstuff()
 		<-e.core.hotstuff.Done()
 		e.log.Debug().Msg("all components have been shut down")
```
Comment (zhangchiqing) [inline engine/consensus/compliance/engine.go:193]:
```suggestion
		e.log.Info().Msg("shutting down hotstuff eventloop")
		e.stopHotstuff()
		<-e.core.hotstuff.Done()
		e.log.Info().Msg("all components have been shut down")
```

File: engine/collection/compliance/engine.go [hunk 5]
```diff
@@ -206,6 +214,7 @@ func (e *Engine) Ready() <-chan struct{} {
 func (e *Engine) Done() <-chan struct{} {
 	e.lm.OnStop(func() {
 		e.log.Debug().Msg("shutting down hotstuff eventloop")
+		e.stopHotstuff()
 		<-e.core.hotstuff.Done()
 		e.log.Debug().Msg("all components have been shut down")
```
Comment (zhangchiqing) [inline engine/collection/compliance/engine.go:219]:
```suggestion
		e.log.Info().Msg("shutting down hotstuff eventloop")
		e.stopHotstuff()
		<-e.core.hotstuff.Done()
		e.log.Info().Msg("all components have been shut down")
```

File: consensus/hotstuff/eventloop/event_loop.go [hunk 6]
```diff
@@ -26,16 +26,16 @@ type EventLoopV2 struct {
 	startTime          time.Time
 }
 
-var _ hotstuff.EventLoopV2 = &EventLoopV2{}
-var _ module.ReadyDoneAware = &EventLoopV2{}
-var _ module.Startable = &EventLoopV2{}
+var _ hotstuff.EventLoop = &EventLoop{}
+var _ module.ReadyDoneAware = &EventLoop{}
+var _ module.Startable = &EventLoop{}
```
Comment (zhangchiqing) [inline consensus/hotstuff/eventloop/event_loop.go:31]:
```suggestion
var _ hotstuff.EventLoop = (*EventLoop)(nil)
var _ module.ReadyDoneAware = (*EventLoop)(nil)
var _ module.Startable = (*EventLoop)(nil)
```

**Review by zhangchiqing — APPROVED** [review]
Looks good

File: consensus/hotstuff/event_loop.go [hunk 7]
```diff
@@ -1,216 +1,13 @@
 package hotstuff
 
-import (
-	"time"
+import "github.com/onflow/flow-go/model/flow"
 
-	"github.com/rs/zerolog"
+// EventLoop performs buffer and processing of incoming proposals and QCs.
+type EventLoop interface {
```
Comment (AlexHentschel) [inline consensus/hotstuff/event_loop.go:6]: I am wondering why we need this interface. In the following code, you have replaced the `EventLoop` with `module.HotSuff`, which seems right to me ... 
Maybe the difference is that `EventLoop` has the additional method `SubmitTrustedQC` which HotStuff doesn't have? If that is the case, we could write

```golang
type EventLoop interface {
	module.HotStuff
	
	// SubmitTrustedQC accepts QC for processing. QC will be dispatched on worker thread.
	// CAUTION: QC is trusted (_not_ validated again)
	SubmitTrustedQC(qc *flow.QuorumCertificate)
}
```

File: consensus/participant.go [hunk 8]
```diff
@@ -41,7 +42,7 @@ func NewParticipant(
 	finalized *flow.Header,
 	pending []*flow.Header,
 	options ...Option,
-) (*hotstuff.EventLoop, error) {
+) (module.HotStuff, error) {
```
Comment (AlexHentschel) [inline consensus/participant.go:45]: this is one of those cases, where I thought: "Why do we need the `hotstuff.EventLoop` interfaces, when we are essentially not using it and only have a single implementation?"

File: engine/collection/compliance/engine.go [hunk 9]
```diff
@@ -195,6 +198,11 @@ func (e *Engine) Ready() <-chan struct{} {
 	}
 	e.lm.OnStart(func() {
 		e.unit.Launch(e.loop)
```
Comment (AlexHentschel) [inline engine/collection/compliance/engine.go:200]: Would be great to also update the Collectors' Compliance Engine to the new `ComponentManager` ... but lets do that in a separate PR

File: engine/consensus/compliance/engine.go [hunk 10]
```diff
@@ -168,7 +171,13 @@ func (e *Engine) Ready() <-chan struct{} {
 	}
 	e.lm.OnStart(func() {
 		e.unit.Launch(e.loop)
```
Comment (AlexHentschel) [inline engine/consensus/compliance/engine.go:173]: refactoring the code to comply with the new convention for `ReadyDoneAware` would be great; probably as a separate PR; using `ComponentManager` would be great

**Review by AlexHentschel — APPROVED** [review]
Thanks for splitting up the code. That really helped clarity. 
LGTM.

File: consensus/hotstuff/event_loop.go (hunk 7 above)
Comment (durkmurder) [reply consensus/hotstuff/event_loop.go:6]: That's one of the reasons. Originally I have added interface to make it uniform with other components. Agree, let's do it this way.

File: engine/collection/compliance/engine.go (hunk 9 above)
Comment (durkmurder) [reply engine/collection/compliance/engine.go:200]: I wanted to do that but:
1) It increases scope of PR a lot
2) We depend on some unit functionality which is not present in component manager.
